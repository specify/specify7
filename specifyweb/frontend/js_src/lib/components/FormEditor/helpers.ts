import type { RR } from '../../utils/types';
import type { Tables } from '../DataModel/types';

export const tablesWithBusRulesIn6 = new Set([
  'AccessionAgentBusRules',
  'AccessionAuthorizationBusRules',
  'AccessionBusRules',
  'AddressBusRules',
  'AgentBusRules',
  'AppraisalBusRules',
  'AttachmentBusRules',
  'AuthorBusRules',
  'BioStratBusRules',
  'BorrowBusRules',
  'CatalogNumberingSchemeBusRules',
  'CatAutoNumberingSchemeBusRules',
  'CollectingEventBusRules',
  'CollectingTripAuthorizationBusRules',
  'CollectingTripBusRules',
  'CollectionBusRules',
  'CollectionObjectBusRules',
  'CollectorBusRules',
  'ConservDescriptionBusRules',
  'ConservEventBusRules',
  'ContainerBusRules',
  'DataTypeBusRules',
  'DeaccessionAgentBusRules',
  'DeterminationBusRules',
  'DeterminerBusRules',
  'DisciplineBusRules',
  'DivisionBusRules',
  'DNASequenceBusRules',
  'DNASequencingRunBusRules',
  'ExchangeOutBusRules',
  'ExchangeOutPrepBusRules',
  'ExtractorBusRules',
  'FieldNotebookBusRules',
  'FieldNotebookPageSetBusRules',
  'FundingAgentBusRules',
  'GeographyBusRules',
  'GeologicTimePeriodBusRules',
  'GiftBusRules',
  'GiftPreparationBusRules',
  'GroupPersonBusRules',
  'InfoRequestBusRules',
  'LithoStratBusRules',
  'LoanBusRules',
  'LoanGiftShipmentBusRules',
  'LoanPreparationBusRules',
  'LoanReturnPreparationBusRules',
  'LocalityBusRules',
  'PcrPersonBusRules',
  'PermitBusRules',
  'PickListBusRules',
  'PreparationBusRules',
  'PrepTypeBusRules',
  'SpecifyUserBusRules',
  'StorageBusRules',
  'TaxonBusRules',
  'TreatmentEventBusRules',
  'TreeDefBusRules',
]);

/**
 * Most of the time business rules class name can be inferred from table name.
 * Exceptions:
 */
export const businessRulesOverride: Partial<RR<keyof Tables, string>> = {
  Shipment: 'LoanGiftShipment',
  CollectionObjectAttachment: 'Attachment',
};

export const getBusinessRuleClassFromTable = (
  tableName: keyof Tables
): string =>
  businessRulesOverride[tableName] === undefined
    ? tablesWithBusRulesIn6.has(`${tableName}BusRules`)
      ? `edu.ku.brc.specify.datamodel.busrules.${tableName}BusRules`
      : ''
    : `edu.ku.brc.specify.datamodel.busrules.${businessRulesOverride[
        tableName
      ]!}BusRules`;
