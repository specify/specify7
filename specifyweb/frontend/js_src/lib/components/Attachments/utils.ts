import { f } from '../../utils/functools';
import { filterArray } from '../../utils/types';
import type { Relationship } from '../DataModel/specifyField';
import type { SpecifyTable } from '../DataModel/specifyTable';
import { genericTables, getTable } from '../DataModel/tables';
import { hasTablePermission } from '../Permissions/helpers';

export const attachmentRelatedTables = f.store(() =>
  Object.keys(genericTables).filter((tableName) =>
    tableName.endsWith('Attachment')
  )
);

export const allTablesWithAttachments = f.store(() =>
  filterArray(
    attachmentRelatedTables().map((tableName) =>
      getTable(tableName.slice(0, -1 * 'Attachment'.length))
    )
  )
);

/** Exclude tables without read access */
export const tablesWithAttachments = f.store(() =>
  allTablesWithAttachments().filter((model) =>
    hasTablePermission(model.name, 'read')
  )
);

/**
 * Gets the relationship to a Table's attachment table.
 * e.g., the collectionObjectAttachments relationship from CollectionObject to CollectionObjectAttachments
 * In most cases the relationship is called tableNameAttachments, although fetching the relationship
 * following that schema will not always be consitent as the following models relationship to their
 * attachment table is simply called `attachments`:
 * - DNASequence
 * - FieldNotebook
 * - FieldNotebookPage
 * - FieldNotebookPageSet
 */
export const getAttachmentRelationship = (
  table: SpecifyTable
): Relationship | undefined => {
  if (table.name === 'Attachment') return undefined;
  const commonRelationship = table.field[
    `${table.name[0].toLowerCase()}${table.name.slice(1)}Attachments`
  ] as Relationship | undefined;

  return commonRelationship === undefined
    ? table.relationships.find((relationship) => {
        const relatedModel = relationship.relatedTable.name;
        return (
          relatedModel !== 'Attachment' && relatedModel.endsWith('Attachment')
        );
      })
    : commonRelationship;
};
