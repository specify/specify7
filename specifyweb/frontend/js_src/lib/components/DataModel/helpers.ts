import { f } from '../../utils/functools';
import type { RA, ValueOf } from '../../utils/types';
import { isTreeResource } from '../InitialContext/treeRanks';
import { relationshipIsToMany } from '../WbPlanView/mappingHelpers';
import type {
  AnySchema,
  AnyTree,
  SerializedResource,
  TableFields,
} from './helperTypes';
import type { SpecifyResource } from './legacyTypes';
import type { LiteralField, Relationship } from './specifyField';
import type { SpecifyTable } from './specifyTable';
import type { Tables } from './types';
/**
 * Lookups for operaters and relationships in Queries on the backend are
 * separated by `__`
 * The api supports the same syntax in query paramters.
 * For example `/api/specify/collectionobject/?collection__discipline__name__startswith="Invert"&catalognumber__gt=000000100`
 * fetches all collectionobjects in disciplines which names start with
 * "Invert" with catalognumbers greater than 100
 */
export const djangoLookupSeparator = '__';

export const backboneFieldSeparator = '.';

/**
 * Formats a relationship lookup which is used in a query passed to the backend.
 * For example `formatRelationshipPath('collection', 'discipline', 'division')`
 * becomes `'collection__discipline__division'`
 */
export const formatRelationshipPath = (...fields: RA<string>): string =>
  fields.join(djangoLookupSeparator);

const weekDayMap = {
  Sunday: 1,
  Monday: 2,
  Tuesday: 3,
  Wednesday: 4,
  Thursday: 5,
  Friday: 6,
  Saturday: 7,
};

/**
 * Use this to construct a query using a lookup for Django.
 * Returns an object which can be used as a filter when fetched from the backend.
 * Example: backendFilter('number1').isIn([1, 2, 3]) is the equivalent
 * of {number1__in: [1, 2, 3].join(',')}
 *
 * See the Django docs at:
 * https://docs.djangoproject.com/en/3.2/ref/models/querysets/#field-lookups
 */
export const backendFilter = (field: string) => ({
  equals: (value: number | string) => ({
    [[field, 'exact'].join(djangoLookupSeparator)]: value,
  }),
  contains: (value: string) => ({
    [[field, 'contains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveContains: (value: string) => ({
    [[field, 'icontains'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveStartsWith: (value: string) => ({
    [[field, 'istartswith'].join(djangoLookupSeparator)]: value,
  }),
  startsWith: (value: string) => ({
    [[field, 'startswith'].join(djangoLookupSeparator)]: value,
  }),
  caseInsensitiveEndsWith: (value: string) => ({
    [[field, 'iendswith'].join(djangoLookupSeparator)]: value,
  }),
  endsWith: (value: string) => ({
    [[field, 'endswith'].join(djangoLookupSeparator)]: value,
  }),
  isIn: (value: RA<number | string>) => ({
    [[field, 'in'].join(djangoLookupSeparator)]: value.join(','),
  }),
  isNull: (value: 'false' | 'true' = 'true') => ({
    [[field, 'isnull'].join(djangoLookupSeparator)]: value,
  }),
  greaterThan: (value: number) => ({
    [[field, 'gt'].join(djangoLookupSeparator)]: value,
  }),
  greaterThanOrEqualTo: (value: number) => ({
    [[field, 'gte'].join(djangoLookupSeparator)]: value,
  }),
  lessThan: (value: number) => ({
    [[field, 'lt'].join(djangoLookupSeparator)]: value,
  }),
  lessThanOrEqualTo: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  matchesRegex: (value: string) => ({
    [[field, 'regex'].join(djangoLookupSeparator)]: value,
  }),

  dayEquals: (value: number) => ({
    [[field, 'day'].join(djangoLookupSeparator)]: value,
  }),
  monthEquals: (value: number) => ({
    [[field, 'lte'].join(djangoLookupSeparator)]: value,
  }),
  yearEquals: (value: number) => ({
    [[field, 'year'].join(djangoLookupSeparator)]: value,
  }),
  weekEquals: (value: number) => ({
    [[field, 'week'].join(djangoLookupSeparator)]: value,
  }),
  weekDayEquals: (value: keyof typeof weekDayMap) => ({
    [[field, 'week_day'].join(djangoLookupSeparator)]: weekDayMap[value],
  }),
});

export const isResourceOfType = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
  // @ts-expect-error
): resource is SpecifyResource<Tables[TABLE_NAME]> =>
  resource.specifyTable.name === tableName;

export const toTable = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableName: TABLE_NAME
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  resource.specifyTable.name === tableName
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

export const toResource = <TABLE_NAME extends keyof Tables>(
  resource: SerializedResource<AnySchema>,
  tableName: TABLE_NAME
): SerializedResource<Tables[TABLE_NAME]> | undefined =>
  resource._tableName === tableName
    ? (resource as SerializedResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * The table.field has a very broad type to reduce type conflicts in components
 * that deal with generic schemas (accept AnySchema or a SCHEMA extends AnySchema)
 */
export const getField = <
  SCHEMA extends ValueOf<Tables>,
  FIELD extends TableFields<SCHEMA>
>(
  table: SpecifyTable<SCHEMA>,
  name: FIELD
): FIELD extends keyof SCHEMA['fields'] ? LiteralField : Relationship =>
  table.field[name] as FIELD extends keyof SCHEMA['fields']
    ? LiteralField
    : Relationship;

export const toTreeTable = (
  resource: SpecifyResource<AnySchema>
): SpecifyResource<AnyTree> | undefined =>
  isTreeResource(resource) ? resource : undefined;

export const toTables = <TABLE_NAME extends keyof Tables>(
  resource: SpecifyResource<AnySchema>,
  tableNames: RA<TABLE_NAME>
): SpecifyResource<Tables[TABLE_NAME]> | undefined =>
  f.includes(tableNames, resource.specifyTable.name)
    ? (resource as SpecifyResource<Tables[TABLE_NAME]>)
    : undefined;

/**
 * Example usage:
 * resource: Collector
 * fields: agent -> lastName
 * Would return [agent, lastName] if agent exists
 *
 */
export async function fetchDistantRelated(
  resource: SpecifyResource<AnySchema>,
  fields: RA<LiteralField | Relationship> | undefined
): Promise<
  | {
      readonly resource: SpecifyResource<AnySchema> | undefined;
      readonly field: LiteralField | Relationship | undefined;
    }
  | undefined
> {
  if (
    Array.isArray(fields) &&
    fields.some(
      (field) =>
        field.isRelationship &&
        relationshipIsToMany(field) &&
        field !== fields.at(-1)
    )
  ) {
    console.error(
      'Can not index inside of a -to-many relationship. Use an aggregator instead'
    );
    return undefined;
  }

  const related =
    fields === undefined || fields.length === 0
      ? resource
      : fields.length === 1
      ? await resource.fetch()
      : await resource.rgetPromise(
          fields
            .slice(0, -1)
            .map(({ name }) => name)
            .join(backboneFieldSeparator)
        );

  const field = fields?.at(-1);
  const relatedResource = related ?? undefined;
  return relatedResource === undefined && field === undefined
    ? undefined
    : {
        resource: relatedResource,
        field,
      };
}
