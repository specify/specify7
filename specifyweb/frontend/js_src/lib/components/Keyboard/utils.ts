import type { LocalizedString } from 'typesafe-i18n';

import type { RA } from '../../utils/types';
import { localized } from '../../utils/types';
import type { KeyboardShortcuts, ModifierKey } from './config';
import { keyboardModifierLocalization, keyboardPlatform } from './config';
import { keyJoinSymbol } from './context';

const localizedKeyJoinSymbol = ' + ';
export const localizeKeyboardShortcut = (shortcut: string): LocalizedString =>
  localized(
    shortcut
      .split(keyJoinSymbol)
      .map((key) => keyboardModifierLocalization[key as ModifierKey] ?? key)
      .join(localizedKeyJoinSymbol)
  );

/**
 * If there is a keyboard shortcut defined for current system, use it
 * (also, if current system explicitly has empty array of shortcuts, use it).
 *
 * Otherwise, use the keyboard shortcut from one of the other platforms if set,
 * but change meta to ctrl and vice versa as necessary.
 */
export function resolvePlatformShortcuts(
  shortcut: KeyboardShortcuts
): RA<string> | undefined {
  if ('platform' in shortcut) return shortcut[keyboardPlatform];
  else if ('other' in shortcut)
    return keyboardPlatform === 'windows'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('windows' in shortcut)
    return keyboardPlatform === 'other'
      ? shortcut.other
      : shortcut.other?.map(replaceCtrlWithMeta);
  else if ('mac' in shortcut) return shortcut.other?.map(replaceMetaWithCtrl);
  else return undefined;
}

const replaceCtrlWithMeta = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'ctrl' ? 'meta' : key))
    .join(keyJoinSymbol);

const replaceMetaWithCtrl = (shortcut: string): string =>
  shortcut
    .split(keyJoinSymbol)
    .map((key) => (key === 'meta' ? 'ctrl' : key))
    .join(keyJoinSymbol);
