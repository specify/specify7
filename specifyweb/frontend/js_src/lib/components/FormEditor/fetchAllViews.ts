import type { LocalizedString } from 'typesafe-i18n';

import { ajax } from '../../utils/ajax';
import type { RA } from '../../utils/types';
import { localized } from '../../utils/types';
import { defined } from '../../utils/types';
import { camelToHuman, split } from '../../utils/utils';
import { strictParseResourceUrl } from '../DataModel/resource';
import type { Tables } from '../DataModel/types';
import type { ViewDefinition } from '../FormParse';
import { userInformation } from '../InitialContext/userInformation';
import { formatUrl } from '../Router/queryString';

type PresentableViewDefinition = ViewDefinition & {
  readonly category: string;
  readonly editUrl: string | undefined;
  readonly disciplineId: number | undefined;
};

export type AllTableViews = {
  readonly database: RA<
    PresentableViewDefinition & { readonly collectionId: number }
  >;
  readonly disk: RA<PresentableViewDefinition>;
};

/**
 * Fetch all views for a given table accessible to current user in each collection
 * Note: this may result in duplicates
 */
export const fetchAllViews = async (
  tableName: keyof Tables,
  cache = false
): Promise<AllTableViews> =>
  Promise.all(
    userInformation.availableCollections.map(async ({ id }) =>
      ajax<RA<ViewDefinition>>(
        formatUrl('/context/views.json', {
          table: tableName,
          collectionId: id,
        }),
        {
          headers: {
            Accept: 'application/json',
          },
          cache: cache ? undefined : 'no-cache',
        }
      ).then(({ data }) => data.map((view) => ({ ...view, collectionId: id })))
    )
  ).then((data) => {
    const [disk, database] = split(
      data.flat(),
      (view) => view.viewsetFile === null
    );
    /*
     * Note, several requests may return the same view definition
     */
    return {
      // Deduplicate views from database
      database: Object.values(
        Object.fromEntries(
          database.map((view) => [`${view.viewsetId ?? ''}_${view.name}`, view])
        )
      ).map((view) => augmentDatabaseView(tableName, view)),
      // Deduplicate views from disk
      disk: Object.values(
        Object.fromEntries(
          disk.map((view) => [`${view.viewsetFile ?? ''}_${view.name}`, view])
        )
      ).map(augmentDiskView),
    };
  });

const augmentDatabaseView = (
  tableName: keyof Tables,
  view: ViewDefinition & { readonly collectionId: number }
): PresentableViewDefinition & { readonly collectionId: number } => ({
  ...view,
  category:
    (view.viewsetLevel === 'Collection'
      ? userInformation.availableCollections.find(
          ({ id }) => id === view.collectionId
        )?.collectionName
      : undefined) ?? camelToHuman(view.viewsetName),
  disciplineId: strictParseResourceUrl(
    defined(
      userInformation.availableCollections.find(
        ({ id }) => id === view.collectionId
      )?.discipline
    )
  )[1],
  editUrl:
    view.viewsetId === null
      ? undefined
      : `/specify/resources/view-set/${view.viewsetId}/${tableName}/${view.name}/`,
});

const augmentDiskView = (view: ViewDefinition): PresentableViewDefinition => ({
  ...view,
  category:
    typeof view.viewsetFile === 'string'
      ? filePathToHuman(view.viewsetFile)
      : camelToHuman(view.viewsetLevel),
  disciplineId: undefined,
  editUrl: undefined,
});

export function filePathToHuman(path: string): LocalizedString {
  const parts = path
    .split('/')
    .filter((part) => part !== '.' && part.length > 0);
  const baseParts = parts.slice(0, -1);
  const fileName = parts.at(-1)!.split('.')[0];
  const result = (
    baseParts.at(-1) === fileName ? baseParts : [...baseParts, fileName]
  )
    .map(camelToHuman)
    .join(' > ');
  return localized(result);
}
