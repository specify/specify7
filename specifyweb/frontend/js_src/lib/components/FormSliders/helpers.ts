import { Relationship } from '../DataModel/specifyField';

export const shouldBeToOne = (
  relationship: Relationship | undefined
): boolean =>
  relationship?.type.includes('-to-many') === false ||
  (relationship?.type.includes('-to-many') === true &&
    !relationship.isDependent() &&
    relationship.getReverse()?.isDependent() === true);
