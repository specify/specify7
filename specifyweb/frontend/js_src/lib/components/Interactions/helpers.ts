import { ajax } from '../../utils/ajax';
import { formData } from '../../utils/ajax/helpers';
import type { RA, RestrictedTuple } from '../../utils/types';
import type { AnyInteractionPreparation } from '../DataModel/helperTypes';
import type { Tables } from '../DataModel/types';

export const interactionPrepTables: RestrictedTuple<
  AnyInteractionPreparation['tableName']
> = [
  'DisposalPreparation',
  'ExchangeInPrep',
  'ExchangeOutPrep',
  'GiftPreparation',
  'LoanPreparation',
];

type ExtractInteraction<T extends string> =
  T extends `${infer Prefix}Prep${string}` ? Prefix : never;

export type InteractionWithPreps = Tables[ExtractInteraction<
  AnyInteractionPreparation['tableName']
>];

export const interactionsWithPrepTables: RestrictedTuple<
  InteractionWithPreps['tableName']
> = ['Disposal', 'ExchangeIn', 'ExchangeOut', 'Gift', 'Loan'];

export type PreparationData = {
  readonly catalogNumber: string;
  readonly collectionObjectId: number;
  readonly taxon: string;
  readonly taxonId: number;
  readonly preparationId: number;
  readonly prepType: string;
  readonly countAmount: number;
  readonly loaned: number;
  readonly gifted: number;
  readonly exchanged: number;
  readonly available: number;
};

export type PreparationRow = readonly [
  catalogNumber: string,
  collectionObjectId: number,
  taxonFullName: string,
  taxonId: number,
  preparationId: number,
  prepType: string,
  preparationCountAmt: number,
  amountLoaned: string | null,
  amountedGifted: string | null,
  amountExchanged: string | null,
  amountAvailable: string
];

export const getPrepsAvailableForLoanRs = async (
  recordSetId: number,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>(
    `/interactions/preparations_available_rs/${recordSetId}/`,
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ isLoan }),
    }
  ).then(({ data }) => data);

export const getPrepsAvailableForLoanCoIds = async (
  idField: string,
  collectionObjectIds: RA<string>,
  isLoan: boolean
) =>
  ajax<RA<PreparationRow>>('/interactions/preparations_available_ids/', {
    method: 'POST',
    headers: { Accept: 'application/json' },
    body: formData({
      id_fld: idField,
      co_ids: collectionObjectIds,
      isLoan,
    }),
  }).then(({ data }) => data);

export const getInteractionsForPrepId = async (prepId: number) =>
  ajax<RA<readonly [number, string | null, string | null, string | null]>>(
    '/interactions/prep_interactions/',
    {
      method: 'POST',
      headers: { Accept: 'application/json' },
      body: formData({ prepIds: prepId }),
    }
  ).then(({ data }) => data[0]);

export const getPrepAvailability = async (
  prepId: number,
  iPrepId: number | undefined,
  iPrepName: string
) =>
  ajax(
    `/interactions/prep_availability/${prepId}/${
      typeof iPrepId === 'number' ? `${iPrepId}/${iPrepName}/` : ''
    }`,
    { headers: { Accept: 'application/json' } }
  ).then(({ data }) => data);
