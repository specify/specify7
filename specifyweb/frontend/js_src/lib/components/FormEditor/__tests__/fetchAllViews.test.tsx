import { requireContext } from '../../../tests/helpers';
import { theories } from '../../../tests/utils';
import { fetchAllViews, filePathToHuman } from '../fetchAllViews';

requireContext();

test('fetchAllViews', async () =>
  expect(fetchAllViews('Accession')).resolves.toMatchSnapshot());

theories(filePathToHuman, [
  {
    in: ['common/common.views.xml'],
    out: 'Common',
  },
  {
    in: ['common/global.views.xml'],
    out: 'Common > Global',
  },
  {
    in: ['common/global/manager.views.xml'],
    out: 'Common > Global > Manager',
  },
  {
    in: ['manager.all.views.xml'],
    out: 'Manager',
  },
  {
    in: ['./common/manager.all.views.xml'],
    out: 'Common > Manager',
  },
  {
    in: ['/common/file/manager.all.views.xml'],
    out: 'Common > File > Manager',
  },
]);
