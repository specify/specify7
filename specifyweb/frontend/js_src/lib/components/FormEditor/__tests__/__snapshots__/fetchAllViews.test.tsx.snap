// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAllViews 1`] = `
{
  "database": [
    {
      "altviews": {
        "Accession Edit": {
          "mode": "edit",
          "name": "Accession Edit",
          "viewdef": "Accession",
        },
        "Accession View": {
          "default": "true",
          "mode": "view",
          "name": "Accession View",
          "viewdef": "Accession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Voucher Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 4,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/4/Accession/Accession/",
      "name": "Accession",
      "resourcelabels": "false",
      "view": "[OMITTED]",
      "viewdefs": {
        "Accession": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 4,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "AccessionCO Edit": {
          "mode": "edit",
          "name": "AccessionCO Edit",
          "viewdef": "AccessionCO",
        },
        "AccessionCO View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionCO View",
          "viewdef": "AccessionCO",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Voucher Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 4,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/4/Accession/AccessionCO/",
      "name": "AccessionCO",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionCO": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 4,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "Accession Edit": {
          "mode": "edit",
          "name": "Accession Edit",
          "viewdef": "Accession",
        },
        "Accession View": {
          "default": "true",
          "mode": "view",
          "name": "Accession View",
          "viewdef": "Accession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Tissue Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 32768,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/6/Accession/Accession/",
      "name": "Accession",
      "resourcelabels": "false",
      "view": "[OMITTED]",
      "viewdefs": {
        "Accession": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 6,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "AccessionCO Edit": {
          "mode": "edit",
          "name": "AccessionCO Edit",
          "viewdef": "AccessionCO",
        },
        "AccessionCO View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionCO View",
          "viewdef": "AccessionCO",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Tissue Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 32768,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/6/Accession/AccessionCO/",
      "name": "AccessionCO",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionCO": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 6,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "Accession Edit": {
          "mode": "edit",
          "name": "Accession Edit",
          "viewdef": "Accession",
        },
        "Accession View": {
          "default": "true",
          "mode": "view",
          "name": "Accession View",
          "viewdef": "Accession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Teaching Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 65536,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/5/Accession/Accession/",
      "name": "Accession",
      "resourcelabels": "false",
      "view": "[OMITTED]",
      "viewdefs": {
        "Accession": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 5,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "AccessionCO Edit": {
          "mode": "edit",
          "name": "AccessionCO Edit",
          "viewdef": "AccessionCO",
        },
        "AccessionCO View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionCO View",
          "viewdef": "AccessionCO",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Teaching Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 65536,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/5/Accession/AccessionCO/",
      "name": "AccessionCO",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionCO": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 5,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "Accession Edit": {
          "mode": "edit",
          "name": "Accession Edit",
          "viewdef": "Accession",
        },
        "Accession View": {
          "default": "true",
          "mode": "view",
          "name": "Accession View",
          "viewdef": "Accession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Observation Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/7/Accession/Accession/",
      "name": "Accession",
      "resourcelabels": "false",
      "view": "[OMITTED]",
      "viewdefs": {
        "Accession": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 7,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
    {
      "altviews": {
        "AccessionCO Edit": {
          "mode": "edit",
          "name": "AccessionCO Edit",
          "viewdef": "AccessionCO",
        },
        "AccessionCO View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionCO View",
          "viewdef": "AccessionCO",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "KU Fish Observation Collection",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": 3,
      "editUrl": "/specify/resources/view-set/7/Accession/AccessionCO/",
      "name": "AccessionCO",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionCO": "[OMITTED]",
      },
      "viewsetFile": null,
      "viewsetId": 7,
      "viewsetLevel": "Collection",
      "viewsetName": "Fish Views",
      "viewsetSource": "db",
    },
  ],
  "disk": [
    {
      "altviews": {
        "Accession Edit": {
          "mode": "edit",
          "name": "Accession Edit",
          "viewdef": "Accession",
        },
        "Accession View": {
          "default": "true",
          "mode": "view",
          "name": "Accession View",
          "viewdef": "Accession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "Common",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": undefined,
      "editUrl": undefined,
      "name": "Accession",
      "view": "[OMITTED]",
      "viewdefs": {
        "Accession": "[OMITTED]",
      },
      "viewsetFile": "common/common.views.xml",
      "viewsetId": null,
      "viewsetLevel": "Common",
      "viewsetName": "Common",
      "viewsetSource": "disk",
    },
    {
      "altviews": {
        "AccessionCO Edit": {
          "mode": "edit",
          "name": "AccessionCO Edit",
          "viewdef": "AccessionCO",
        },
        "AccessionCO View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionCO View",
          "viewdef": "AccessionCO",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "Common",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": undefined,
      "editUrl": undefined,
      "name": "AccessionCO",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionCO": "[OMITTED]",
      },
      "viewsetFile": "common/common.views.xml",
      "viewsetId": null,
      "viewsetLevel": "Common",
      "viewsetName": "Common",
      "viewsetSource": "disk",
    },
    {
      "altviews": {
        "AccessionDeaccession Table View": {
          "default": "true",
          "mode": "view",
          "name": "AccessionDeaccession Table View",
          "viewdef": "AccessionDeaccession Table",
        },
        "AccessionDeaccession View": {
          "mode": "view",
          "name": "AccessionDeaccession View",
          "viewdef": "AccessionDeaccession",
        },
      },
      "busrules": "edu.ku.brc.specify.datamodel.busrules.AccessionBusRules",
      "category": "Common",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": undefined,
      "editUrl": undefined,
      "name": "AccessionDeaccession",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionDeaccession": "[OMITTED]",
        "AccessionDeaccession Table": "[OMITTED]",
      },
      "viewsetFile": "common/common.views.xml",
      "viewsetId": null,
      "viewsetLevel": "Common",
      "viewsetName": "Common",
      "viewsetSource": "disk",
    },
    {
      "altviews": {
        "AccessionSearch": {
          "default": "true",
          "mode": "edit",
          "name": "AccessionSearch",
          "validated": "false",
          "viewdef": "AccessionSearch",
        },
      },
      "category": "Backstop > Search",
      "class": "edu.ku.brc.specify.datamodel.Accession",
      "collectionId": 98304,
      "disciplineId": undefined,
      "editUrl": undefined,
      "name": "AccessionSearch",
      "resourcelabels": "false",
      "usedefbusrule": "false",
      "view": "[OMITTED]",
      "viewdefs": {
        "AccessionSearch": "[OMITTED]",
      },
      "viewsetFile": "backstop/search.views.xml",
      "viewsetId": null,
      "viewsetLevel": "Backstop",
      "viewsetName": "Search",
      "viewsetSource": "disk",
    },
  ],
}
`;
